lib/Conversion/LinalgToArith/LinalgToArith.cpp

```diff
@@ -0,0 +1,385 @@
+    Secretness leftOperandSecretness =
+        leftOperandLookup ? leftOperandLookup->getValue() : Secretness();
+    bool leftOperandIsSecret = (leftOperandSecretness.isInitialized() &&
+                                leftOperandSecretness.getSecretness());
+
+    auto *rightOperandLookup =
+        solver->lookupState<SecretnessLattice>(op.getInputs()[1]);
+    Secretness rightOperandSecretness =
+        rightOperandLookup ? rightOperandLookup->getValue() : Secretness();
+    bool rightOperandIsSecret = (rightOperandSecretness.isInitialized() &&
+                                 rightOperandSecretness.getSecretness());
+
+    LLVM_DEBUG({
+      llvm::dbgs() << "Left operand is secret: " << leftOperandIsSecret << "\n"
+                   << "Right operand is secret: " << rightOperandIsSecret
+                   << "\n";
+    });
+
+    // Error out if both are secret or both are public
+    if ((leftOperandIsSecret && rightOperandIsSecret) ||
+        (!leftOperandIsSecret && !rightOperandIsSecret)) {
+      return failure();
+    }
+    auto inputs = op.getInputs();
+    auto outputs = op.getOutputs();
+
+    Value secretValues;
+    Value publicMatrix;
+    if (leftOperandIsSecret) {
+      secretValues = inputs[0];
+      publicMatrix = inputs[1];
+    } else {  // the right operand is secret
+      secretValues = inputs[1];
+      publicMatrix = inputs[0];
+    }
+    auto matrixTensorType = cast<RankedTensorType>(publicMatrix.getType());
+    auto bias = outputs[0];
+
+    auto dimensions = matrixTensorType.getShape();
+    int64_t dim0 = dimensions[0];  // This is the number of rows in the matrix.
+    int64_t dim1 = dimensions[1];  // This is the number of columns in the
+    // matrix.
+
+    SmallVector<int64_t> transposedDimensions({dim1, dim0});
+
+    // If one of these dimensions is not a power of two, then we can't do the
+    // Halevi-Shoup or Squat Packing Matrix Multiplication conversion.
+    if (!isPowerOfTwo(dim0) || !isPowerOfTwo(dim1)) {
+      return failure();
+    }
+
+    // If the matrix is not a square matrix, then we are doing squat packing.
+    // If so, then the vector dimension is also the longer dimension.
+    if ((leftOperandIsSecret && dim0 < dim1) ||
+        (rightOperandIsSecret && dim0 > dim1)) {
+      return failure();
+    }
+
+    // Diagonalize the matrix first.
+    auto constantValues =
+        dyn_cast<arith::ConstantOp>(publicMatrix.getDefiningOp());
+    if (!constantValues) {
+      return failure();
+    }
+
+    DenseElementsAttr denseAttr =
+        dyn_cast<DenseElementsAttr>(constantValues.getValueAttr());
+
+    // If the constant values doesn't have a dense attribute, then we can't
+    // diagonalize the matrix.
+    if (!denseAttr) {
+      return failure();
+    }
+
+    auto type = denseAttr.getElementType();
+    bool intType = llvm::isa<IntegerType>(type);
+    bool floatType = llvm::isa<FloatType>(type);
+    if (!intType && !floatType) {
+      return failure();
+    }
```

Comment:
I think you should be able to use https://github.com/llvm/llvm-project/blob/eec1fac9b51d06c8afafe9952a20ba7cd4c3ce1c/mlir/lib/IR/Types.cpp#L121

```suggestion
    if (!denseAttr.getElementType().isIntOrFloat()) {
      return failure();
    }
```